File: lc/520/q4.cpp
#include <algorithm>
#include <vector>
#include <iostream>
using std::vector;
template <class T>
concept IsVector = requires(T t) {
    []<class U, class A>(const std::vector<U, A>&){}(t);
};

template <class T>
void print(T a) {
	if constexpr (IsVector<T>) {
		if (a.size() == 0) {
			std::cout << "[]";
		} else {
			std::cout << "[";
			print(a[0]);
			for (int i = 1; i < a.size(); i++) {
				print(",");
				print(a[i]);
			}
			std::cout << "]";
		}
	} else {
		std::cout << a;
	}
}

class Solution {
public:
    vector<int> largestPower(vector<int>& nums) {
		vector<vector<int>> seg;
		int n = nums.size();

		seg.push_back(nums);
		int M = 14;
		vector<int> L;
		vector<int> power(15);
		int mx = *std::max_element(nums.begin(), nums.end());
		M = std::__lg(mx);
		while (M >= 0) {
			int pre = 0;
			for (int i = 0; i < seg.size(); i++) {
				vector<int> one, zero;
				for (int x: seg[i]) {
					if (x >> M & 1) {
						one.push_back(x);
					} else {
						zero.push_back(x);
					}
				}
				pre += one.size();
				if (zero.size() == 0) continue;
				if (one.size() > 0) {
					seg[i] = std::move(one);
					seg.insert(seg.begin() + i + 1, zero);
				}
			}
			power[14 - M] = pre;

		}
		return power;
    }
};


int main() {
	Solution s;
	vector<int> a = {1, 4};
	vector<int> ans = s.largestPower(a);

	for (int i = 0; i < ans.size(); i++) {
		std::cout << ans[i] << " \n"[i == ans.size() - 1];
	}
}

File: lc/520/q4.py
class Trie01:
    def __init__(self, n):
        self.t = [0] * (n + 1)
        self.l = [0] * (n + 1)
        self.r = [0] * (n + 1)
        self.cur = 0
        self.D = 14

    def newNode(self):
        self.cur += 1
        return self.cur

    def insert(self, x):
        u = 0
        d = self.D
        while d >= 0:
            if x >> d & 1 == 1:
                if self.r[u] == 0:
                    self.r[u] = self.newNode()
                u = self.r[u]
            else:
                if self.l[u] == 0:
                    self.l[u] = self.newNode()
                u = self.l[u]
            self.t[u] += 1
            d -= 1
        return u

    def ans(self):
        power = [0] * (15)
        # print(self.t)
        u = 0
        d = self.D
        while d >= 0:
            while d >= 0 and self.r[u] == 0:
                u = self.l[u]
                d -= 1
            u = self.r[u]
            power[14 - d] = self.t[u]
            d -= 1
        return power


class Solution:
    def largestPower(self, nums: list[int]) -> list[int]:
        n = len(nums)
        x = max(nums)

        l = []
        r = list(nums)
        power = [0] * 15
        pre = 0
        while x > 0:
            d = x.bit_length() - 1
            ll = []
            rr = []
            xx = x
            for y in r:
                if y >> d & 1 == 1:
                    rr.append(y)
                    xx &= y
                else:
                    ll.append(y)
            for i in range(len(rr)):
                rr[i] ^= x
            rr.sort(key=lambda x: -x)
            xxx = xx ^ x

            power[14 - d] = len(rr)


# 最大放最左侧，
# 然后枚举这个最大的若干位，

s = Solution()
print(s.largestPower([7, 5]))
print(s.largestPower([3, 1, 7]))
